test/dev/dev.test.ts

```diff
@@ -0,0 +1,52 @@
+const path = require('path');
+const Stream = require('stream');
+
+const execa = require('execa');
+const {readdirSync, readFileSync, statSync, existsSync} = require('fs');
+const glob = require('glob');
+const os = require('os');
+const got = require('got');
+
+describe('snowpack dev', () => {
+  it('smoke', async () => {
+    const cwd = path.join(__dirname, 'smoke');
+    const streamLogsToOutput = new Stream.Writable({objectMode: true});
+
+    // start the server
+    const snowpackProcess = execa('yarn', ['teststart'], {cwd});
+    snowpackProcess.stdout.pipe(streamLogsToOutput);
+
+    // await server to be ready and set a timeout in case something goes wrong
+    await new Promise((resolve, reject) => {
+      // start timeout in case something goes wrong.
+      const timeout = setTimeout(() => {
+        snowpackProcess.cancel();
+        console.error(output.join(''));
+        reject(new Error('Timeout: snowpack did not start server within 3 seconds.'));
+      }, 3000);
+
+      const output = [];
+      streamLogsToOutput._write = (buffer, encoding, done) => {
+        const line = buffer.toString();
+        output.push(line);
+        if (/Server started in/.test(line)) {
+          resolve();
+          clearTimeout(timeout);
+        }
+
+        done();
+      };
+    });
+
+    // get HTML
+    const {body: htmlBody} = await got('http://localhost:8080');
+    expect(htmlBody).toMatchSnapshot('html');
+
+    // get built JS
+    const {body: jsBody} = await got('http://localhost:8080/_dist_/index.js');
+    expect(jsBody).toMatchSnapshot('js');
```

Comment:
In order to make the tests work with any kinds of setups, we could use something like [website-scraper](https://www.npmjs.com/package/website-scraper) to download all assets linked from the HTML page and compare it to the snapshot. That would allow us to have a similar setup to what we currently have in `test/build/*`

Reply:
Not a request for change in this PR, but Puppeteer can do this fairly-easily. Something like this is all that’s needed:

```js
const puppeteer = require('puppeteer');

let browser;

describe('snowpack dev', () => {
  beforeAll(async () => {
    browser = await puppeteer.launch();
  });

  afterAll(async () => {
    await browser.close();
  });

  it('test', async () => {
    const page = await browser.newPage();
    page.on('response', async (res) => {
      if (res.status() === 200) {
        const contents = await res.text();  // note: may need some sanitation / normalization
        const id = res.url();
        expect(contents).toMatchSnapshot(id);
      }
    });
    await page.goto('localhost:8080', {waitUntil: 'networkidle0'});
  })
});
```

Reply:
I think puppeteer is a pretty heavy dependency to have, I'm hoping for something more lightweight. 

On the other side we might introduce puppeteer or playwright anyway as dependency if we want to do end-to-end tests for the templates. I'd say let's do a follow up discussion for it

Reply:
Yeah. +1 for having a discussion.

I did look at the project linked above 👆🏻, but it says it doesn’t execute JS to pull more imports. If we wanted that (which we probably would, to trace ESM imports), I don’t think that’s possible w/o Puppeteer or something equivalent that ships an engine. But that’s only if we wanted our test suite to be more “automatic.”

We could easily do something lightweight if we manually declare all the URLs instead, which may have other advantages as well.